### DISCUSSION TITLE
Can i access C #define's in nelua code?

### DISCUSSION THREAD

**Discussion by ghost** [edubart/nelua-lang#128] [Q&A] [answered]
If i do something like this( i know ccinfo.is_winodws exists )
```lua
##[[ 
if _WIN32 then 
      -- do some thing
end
]]
```
is that looking for a variable named as ``_WIN32`` in the nelua pre-processor code?
or
does it also look in the GCC/OS?, so if i don't have a variable named as ``_WIN32`` in the nelua pre-processor code, it will still work because _WIN32 is defined in GCC or in the OS

**Comment by edubart** [comment, marked as answer]
> Is that looking for a variable named as _WIN32 in the nelua pre-processor code?

When you use an `if` statement in the preprocessor, it will look for a variable in the preprocessor's Lua environment, and `_WIN32` is not defined there, unless you define it (by doing `_WIN32 = true` before the if).

> Does it also look in the GCC/OS?

No, the Nelua's preprocessor does not make any define from the C preprocessor visible, but some from the C compiler are converted to values you can find in `ccinfo`, so you can do `if ccinfo.is_windows then` to check for Windows, under the hood this will be true in the C compiler also defines `_WIN32`, and it's an effective way to detect windows.

> Can I access C #define's in Nelua code?

I'm going to explain the current ways to do this below.

## 1. Using `ccinfo` (recommended)

This is the recommended way to use values present in `ccinfo`. The variable `ccinfo` is a table containing information about the target features and C compiler features, you should use it to detect different OS, architectures, C compilers.

```lua
-- Using ccinfo (recommended)
## if ccinfo.is_win32 then
  print 'target is win32'
## end
```

## 2. Injecting C `#ifdef` (not recommended)

Injecting C ifdefs is also possible, this can be useful if you want to generate portable C code, by portable I mean a C code that can be compiled in different systems. However this approach is not recommended, because this will not work with some language features, like `defer` and `<close>` variables, because this hides the control flow from the Nelua compiler.

```lua
-- Injecting C ifdefs (not recommended)
## cemit '#ifdef _WIN32'
  print 'target is win32'
## cemit '#endif'
```

## 3. Parsing C defines using the C compiler

Nelua has a utility to parse C defines from any C header into a table containing all visible C defines. I don't encourage doing this because it will slowdown compilation due to extra process executions, and the table may contain values you would not expect compiling with a different C compiler or target.

```lua
-- Parsing C defines with your C compiler
## local ccdefs = require'nelua.ccompiler'.get_cc_defines('<windows.h>')
## if ccdefs._WIN32 then
  print 'target is win32'
  print('WINVER ', #[ccdefs.WINVER]#)
  print('_WIN32_WINNT', #[ccdefs._WIN32_WINNT]#)
## end
```

## 4. Setting defines manually

You can define preprocessor variables manually, this is the most simple, but requires user configuration, defines can also be passed with command line arguments, for example with `nelua -DPLATFORM_WINDOWS myprog.lua`.

```lua
-- Setting your define manually
## PLATFORM_WINDOWS = true
## if PLATFORM_WINDOWS then
  print 'target is windows platform!'
## end
```

**Reply by ghost** [reply]
Okay, Thank you so much :+1:
